proposals/0759-react-native-frameworks.md

```diff
@@ -0,0 +1,454 @@
+- Work with existing React Native Framework to identify use cases that are not properly covered by their Framework yet
+- Work with existing React Native Framework to provide a viable migration path from the template.
+
+##### Developers that can’t migrate to an existing React Native Framework because it doesn’t satisfy their requirements
+
+We do have several developers who can’t use a Framework yet, due to specific constraints.
+
+One of the partners that we had conversation with, is developing React Native apps inside a monorepo with multiple native apps (a mixture of RN-only and mixed native apps), as well as multiple React web applications. They built their own bespoke Framework, but still rely on the template & upgrade helper as a reference.
+
+For those developers we will:
+
+- Move the current template **outside** of the `react-native` package:
+  - Specifically move it to `@react-native-community/template` and kept on being **maintained by the community and the React Native team**.
+    - This is the least disruptive change for developers, as it will keep on being easily accessible but it will be clearer that is a community owned component, and not part of the core of react-native
+    - Moreover, the template will be closer to where the community CLI lives today (`@react-native-community/cli`)
+  - We (Meta & the community) want to keep on **maintaining** the template and make sure that developers that can’t migrate to a framework, still have a viable alternative.
+  - We don’t foresee the template to change drastically in the future. Only **minimal changes** will be allowed to the template, to reduce the upgrade pain between versions.
+  - Similarly, we’ll aim at **reducing the breaking changes** as much as possible.
+- Update all the various community tools to refer to the new template location (e.g. the Upgrade Helper)
+- Work with existing React Native Framework to identify use cases that are not properly covered by their framework
+
+#### Community CLI
+
+The React Native Community CLI is tightly coupled to the template. Apps like RN-Tester (and the proposed HelloWorld app) are using the community CLI to build and being tested on debug/release on CI. Similarly the community CLI is used to do release testing.
+
+On top of this, having the community CLI living in a separate repository (`@react-native-community/cli`) introduces a number of **circular dependencies** and friction as part of the release process.
+
+Similar to the Template, we envision having a **private CLI** inside the react-native repository to support the internal use cases mentioned above. This will be achieved by evolving each command of the CLI as follows:
+
+##### Commands to be moved inside react-native core
+
+Commands that live inside `react-native` should be extensible enough to be **used by Frameworks or libraries, **if they wish to**.**
+
+- `clean, build-android, build-ios, run-android, run-ios, log-ios, log-android`
```

Comment:
Does that mean that auto linking currently doesn't rely on some specific behavior of the build command? Or is there some additional work required to decouple that? From what I understand, community CLI supports different build configurations including support for out-of-tree platforms (i.e. tvOS).

What I think is the plan, but it isn't clearly stated here, is for `react-native` command to be owned/maintained/published by React Native core team. If this is the case, when you later mention the command should be removed, do you only indent for the implementation to be removed from the core repo, or also making the whole option unavailable when using `react-native` command and move it under a separate community cli command.

Reply:
> Does that mean that auto linking currently doesn't rely on some specific behavior of the build command? Or is there some additional work required to decouple that? From what I understand, community CLI supports different build configurations including support for out-of-tree platforms (i.e. tvOS).

There's some small refactoring work expected before moving the commands implementations to the core, because currently we're passing the config directly.

Reply:
> There's some small refactoring work expected before moving the commands implementations to the core, because currently we're passing the config directly.

Yup basically this. Currently autolinking relies on invoking `react-native config` so that dependencies & project configs are loaded correctly.
We'll have to decouple it so that autolinking just expects an input that framework developers can provide.

> is for react-native command to be owned/maintained/published by React Native core team

That's correct. That's already the case today by the way.

> when you later mention the command should be removed

Are you referring to the `upgrade` & `ram-bundle` command mentioned in the `##### Commands to be removed` section? As for those, we intend the to remove from the CLI entirely so it won't be possible to invoke them at all anymore. Is this consistent with your mental model @thymikee ?